Lib/fontTools/ttLib/removeOverlaps.py

```diff
@@ -0,0 +1,117 @@
+""" Simplify TrueType glyphs by merging overlapping contours/components.
+
+Requires https://github.com/fonttools/skia-pathops
+"""
+
+from typing import Iterable, Optional, Mapping
+
+from fontTools.ttLib import ttFont
+from fontTools.ttLib.tables import _g_l_y_f
+from fontTools.pens.recordingPen import DecomposingRecordingPen
+from fontTools.pens.ttGlyphPen import TTGlyphPen
+
+import pathops
+
+
+_TTGlyphMapping = Mapping[str, ttFont._TTGlyph]
+
+
+def skPathFromSimpleGlyph(glyphName: str, glyphSet: _TTGlyphMapping) -> pathops.Path:
+    path = pathops.Path()
+    pathPen = path.getPen()
+    glyphSet[glyphName].draw(pathPen)
+    return path
+
+
+def skPathFromCompositeGlyph(glyphName: str, glyphSet: _TTGlyphMapping) -> pathops.Path:
+    # record TTGlyph outlines without components
+    dcPen = DecomposingRecordingPen(glyphSet)
+    glyphSet[glyphName].draw(dcPen)
+    # replay recording onto a skia-pathops Path
+    path = pathops.Path()
+    pathPen = path.getPen()
+    dcPen.replay(pathPen)
+    return path
+
+
+def ttfGlyphFromSkPath(path: pathops.Path) -> _g_l_y_f.Glyph:
+    # Skia paths have no 'components', no need for glyphSet
+    ttPen = TTGlyphPen(glyphSet=None)
+    path.draw(ttPen)
+    glyph = ttPen.glyph()
+    assert not glyph.isComposite()
+    # compute glyph.xMin (glyfTable parameter unused for non composites)
+    glyph.recalcBounds(glyfTable=None)
+    return glyph
+
+
+def removeOverlaps(
+    font: ttFont.TTFont, glyphNames: Optional[Iterable[str]] = None
+) -> None:
+    """ Simplify glyphs in TTFont by merging overlapping contours.
+
+    Overlapping components are first decomposed to simple contours, then merged.
+
+    Currently this only works with TrueType fonts with 'glyf' table.
+    Raises NotImplementedError if 'glyf' table is absent.
+
+    Args:
+        font: input TTFont object, modified in place.
+        glyphNames: optional iterable of glyph names (str) to remove overlaps from.
+            By default, all glyphs in the font are processed.
+    """
+    try:
+        glyfTable = font["glyf"]
+    except KeyError:
+        raise NotImplementedError("removeOverlaps currently only works with TTFs")
+
+    hmtxTable = font["hmtx"]
+    glyphSet = font.getGlyphSet()
+
+    if glyphNames is None:
+        glyphNames = font.getGlyphOrder()
+
+    for glyphName in glyphNames:
+        if glyfTable[glyphName].isComposite():
+            path = skPathFromCompositeGlyph(glyphName, glyphSet)
+        else:
+            path = skPathFromSimpleGlyph(glyphName, glyphSet)
+
+        # duplicate path
+        path2 = pathops.Path(path)
+
+        # remove overlaps
+        path2.simplify()
+
+        # replace TTGlyph if simplified copy is different
+        if path2 != path:
```

Comment:
It might be worth adding reporting to indicate if any paths were modified during the script run.

Reply:
maybe we can print the modified glyph names in a logging.debug message, I don't want to pollute the output with too much extra info. Or maybe just an overall stat like the number of simple vs composite glyphs modified in total.

Reply:
Yeah I agree that a dump of all glyphs would be overwhelming.  A summary that it did / did not change something or the total number of glyphs that were changed may be sufficient.  Alternatively, we could add a check to determine if there are any glyphs with overlaps. That will likely be the info that someone needs pre/post script execution.  If not appropriate here, perhaps as a fontbakery check.

Reply:
checking for overlaps isn't trivial and and implies a lot of the work of removing the overlaps themselves. that's why here I first do the overlap removal and then check if the simplified path has changed compared to the original. And also that's why I need to minimize the changes to only those strictly needed by the overlap removal process.
At least for composites, I could optimise this by first checking that the bboxes of each component _don't_ overlap in which case I should avoid the overlap-removal. Maybe later.